lib/Site.js

```diff
@@ -74,6 +77,37 @@ function setExtension(filename, ext) {
   );
 }
 
+/**
+* Creates a function that delays invoking func until after wait milliseconds have elapsed
+* and the previous function have resolved/rejected.
+* @param func the promise-returning function to delay
+* @param wait the number of milliseconds to delay
+*/
+function delay(func, wait) {
+  let context;
+  let pendingArgs = [];
+  let waitingPromise = null;
+  let previousPromise = Promise.resolve();
+
+  return function (...args) {
+    context = this;
+    pendingArgs.push(args);
+
+    if (waitingPromise === null) {
+      waitingPromise = new Promise((resolve, reject) => {
+        Promise.all([Promise.delay(wait), previousPromise]).then(() => {
+          previousPromise = waitingPromise || Promise.resolve();
```

Comment:
This is in a block that checks `if (waitingPromise === null)`.

---
Why don't we just return a chained Promise? Something like:

```js
return waitingPromise.then(() => func.apply(context, pendingArgs));
```

Reply:
Using `rebuildSourceFiles` as an example,
 * `waitingPromise` has not started executing anything in `rebuildSourceFiles`
* `pendingPromise` is still executing  `rebuildSourceFiles` and its promises and is not fulfilled/rejected yet.

Since the args are accumulated, we only need 1 `waitingPromise` (hence the `if (waitingPromise === null)` check)

Before we can do `func.apply(context, pendingArgs)` e.g. rebuild multiple source files, we will have to
1. wait for the previous rebuild to complete (or live reloading will crash)
    * This was previously done with `async-lock`
    * Now, `waitingPromise` must wait for `pendingPromise` to be fulfilled or rejected.
2. maybe wait for some time so we can batch process a few files together
    * waitingPromise wait for `Promise.delay(wait)`
    * This isn't strictly debouncing as it ensures at least k seconds after the _first_ call is made (not the last as in debouncing)

Some examples with wait = 1 sec / further clarification
* Changes are first made to File A. We do not rebuild the file immediately. We wait for 1 second and if File B and C change within this second, they will be rebuilt together.
* If the ongoing rebuild is really slow  (e.g. 2 sec), we will just collect all the subsequent file paths that need to be rebuilt and rebuild them together after the ongoing one is complete (the 1 sec wait was already complete). This is unlike the debounce + lock approach where they will still be done in batches decided by the debounced call.
* Since the handlers in `index.js` expect to get a promise back and chain it, we return the same `waitingPromise` to all calls. If one file is missing which will cause ENOENT, multiple `catch` might be executed.